### PR TITLE
Add JSON-Schema create-params and a common Validate Credentials API

### PR DESCRIPTION
**Pull request by agrare**
Add a set of parameters required by the UI to drive their data-driven-forms and implement the common validate_credentials API instead of raw_connect.

### CODE REVIEW

**Comment by agrare** [comment]
cc @skateman

File: app/models/manageiq/providers/amazon/cloud_manager.rb [hunk 1]
```diff
@@ -119,6 +119,24 @@ def self.default_blacklisted_event_names
     Settings.ems.ems_amazon.blacklisted_event_names
   end
 
+  def self.provider_create_params
+    {
+      self.name => params_for_create
+    }
+  end
```
Comment (skateman) [inline app/models/manageiq/providers/amazon/cloud_manager.rb]: Ok, now I get it, I don't think we actually need this. I can simply work with the one below. Or is there any special reason for this format?

Comment (agrare) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: Nope I was just trying to maintain the same format that would be returned by calling it on the base class.

Comment (skateman) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: Nah, it might make sense to have it on the `ExtManagementSystem` but not here.

Comment (agrare) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: :+1:

File: app/models/manageiq/providers/amazon/manager_mixin.rb [hunk 2]
```diff
@@ -51,6 +51,23 @@ module ClassMethods
...
+    #   "endpoints" => {
+    #     "ec2" => {
+    #       "access_key" => "",
+    #       "secret_access_key" => "",
+    #     }
+    #   }
+    # }
+    def validate_credentials(args)
+      region = args["region"]
+      ec2_endpoint = args.dig("endpoints", "ec2")
```
Comment (Fryguy) [inline app/models/manageiq/providers/amazon/manager_mixin.rb]: dig could return nil if a layer is missing, but the next line doesn't check for nil before going deeper.

**Comment by Fryguy** [comment]
Liking this approach!

File: app/models/manageiq/providers/amazon/manager_mixin.rb (hunk 2 above)
Comment (agrare) [reply app/models/manageiq/providers/amazon/manager_mixin.rb]: :+1: good call

**Comment by skateman** [comment]
I like it too, but RSpec doesn't :confused: what about @hyperkid123?

**Comment by Hyperkid123** [comment]
The schema looks good. Is there supposed to be some validation?

**Comment by skateman** [comment]
@Hyperkid123 that's coming in a separate PR [here](https://github.com/ManageIQ/manageiq/pull/19120), I'll expose it in an API PR soon.

File: app/models/manageiq/providers/amazon/cloud_manager.rb [hunk 3]
```diff
@@ -119,6 +119,18 @@ def self.default_blacklisted_event_names
...
 
+  def self.params_for_create
+    {
+      :title  => "Configure AWS",
+      :fields => [
+        {:component => "text-field", :name => "role",         :type => "hidden", :initialValue => "ec2"},
+        {:component => "text-field", :name => "region",       :label => "Region"},
+        {:component => "text-field", :name => "ec2_username", :label => "Access Key"},
+        {:component => "text-field", :name => "ec2_password", :label => "Secret Key", :type => "password"}
+      ]
```
Comment (skateman) [inline app/models/manageiq/providers/amazon/cloud_manager.rb]: Having this flat structure might not be enough if you're expecting the data in the validation method in a nested form. Here all the items in the DDF schema are on the same level, but below you wrap everything except of `region` around an `endpoint`.

File: app/models/manageiq/providers/amazon/manager_mixin.rb [hunk 4]
```diff
@@ -51,6 +51,25 @@ module ClassMethods
...
+    # args:
+    # {
+    #   "region" => "",
+    #   "endpoints" => {
+    #     "ec2" => {
+    #       "access_key" => "",
+    #       "secret_access_key" => "",
+    #     }
+    #   }
+    # }
```
Comment (skateman) [inline app/models/manageiq/providers/amazon/manager_mixin.rb]: The current DDF schema won't send the data in this form, but:
```ruby
{
  "region" => "",
  "access_key" => "",
  "secret_access_key" => ""
}
```

**Comment by martinpovolny** [comment]
I do not think that we need the schema to be the DDF schema. Rather I'd have something simple that we can use to generate the final DDF schema with some whistles and bells.

Wdyt, @skateman, @Hyperkid123?

File: app/models/manageiq/providers/amazon/cloud_manager.rb (hunk 3 above)
Comment (rvsia) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: For nested values you can write something like `endpoints.ec2.username` and DDF will send it as 
```jsx
endpoints: {
   ec2: {
      username: `myname`
 }
}
```

Comment (agrare) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: Okay great, yeah I'm not very familiar with DDF so was looking to you guys for how to get the structure back that we're looking for.

Comment (agrare) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: Okay how does that look?

Comment (Hyperkid123) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: Nesting like this is great. We can go even step further and add `authentication.default.userid` to match the actual API request payload (if there is any). But we can add prefixes easily later on.

**Comment by Hyperkid123** [comment]
@martinpovolny i am not really against using something different but is there as long as we use something explicit. I would like to avoid having something ambiguous (like the mozzila schema).

We should probably consider if this is purely for UI purposes or is there any other reason why would want this? If there is i would consider thinking about something else. Otherwise i don't see a reason why not to use ddf.

File: app/models/manageiq/providers/amazon/cloud_manager.rb (hunk 3 above)
Comment (agrare) [reply app/models/manageiq/providers/amazon/cloud_manager.rb]: Okay there's only one endpoint so we can just flatten out the structure, good to know how to nest under endpoints for providers like ovirt though :+1:

**Comment by agrare** [comment]
@Hyperkid123 how to do optional parameters in DDF?  I'd like to add proxy-uri and assume-role but they're not required

**Comment by Hyperkid123** [comment]
Do you mean optional in a way that the fields are there but they are not required or if someone chooses some option from select for instance, then another field appears?

If you mean the first case: By default all fields are option (as in they are not required). If you wan to add required you must add validator and additional properties for the component to add visual representation (for required the *).

example: 
```javascript
{
  component: 'text-field',
  name: 'field',
  label: 'Validated field!',
  isRequired: true,
  validate: [
    {
      type: 'required-validator',
    }
  ]
}
```
http://data-driven-forms.surge.sh/renderer/validators#requiredvalidator

**Comment by agrare** [comment]
Oh okay awesome thanks @Hyperkid123

**Comment by Hyperkid123** [comment]
@agrare yes. Currently we have just a limited amount of data types but we can add some if we need: http://data-driven-forms.surge.sh/renderer/data-types

What data-types are we talking about? Or do you want to just check the if the string is URL for instance? For that we have a number of different validators we can use.

**Comment by Hyperkid123** [comment]
@agrare All values are strings bu default. There is no data-type URL in JS, but we can add url validator: http://data-driven-forms.surge.sh/renderer/validators#urlvalidators

There is a lot of params for URL validator like port, host, if ip address is valid, etc. Let me know what url format you want and we can add it.

File: app/models/manageiq/providers/amazon/cloud_manager.rb [hunk 5]
```diff
@@ -119,6 +119,46 @@ def self.default_blacklisted_event_names
...
+          :name      => "endpoints.default.proxy_uri",
+          :label     => "Proxy URI"
+        },
+        {
+          :component => "text-field",
+          :name      => "endpoints.default.assume_role",
+          :label     => "Assume Role"
+        }
+      ]
+    }
```
Comment (Fryguy) [inline app/models/manageiq/providers/amazon/cloud_manager.rb]: Minor, but I feel this should be a constant, which after that I don't have a preference if it's exposed as a method as well and/or just the constant.

**Comment by Fryguy** [comment]
Regarding the validation it would be really sweet if we had a DFF -> Hash validation class on the Ruby side.  Unfortunately, you'd need a Ruby implementation of each validator, but this would get us server-side validation (to help with direct-api calls) in addition to the client-side validation.

**Pull request merged**